### PR TITLE
Phosphorus tracking

### PR DESCRIPTION
**Pull request by militsasotirova**
This pull request has the code for P tracking in the animals. For equations that are common among all animal classes, there are functions in animal_base.py, while equations specific to animal classes are found in the respective file. I set up a new system for adding animals from the replacement herd - a queue is stored in animal_management for which pens have lost animals due to culling or the fact that too many animals exist. Whenever an animal is added, I add that animal to the pen that is at the front of this queue. I also keep a dictionary of which animals are in which pens to make this process easier. Although it shouldn't be the case that an animal is added before an animal is removed, this does occasionally occur, so I added a check before trying to pop from the queue, and if it is empty I add the animal to the last pen (since the animal will be a Cow, so it should go into the last pen). I assume that if the animal needs to be in a different pen, that will be adjusted during the next pen allocation.

**NOTE**: Because P requirements are not currently used in the ration optimization, animal P values may become negative due to inadequate P in the diet. If you'd like to check the changes with a more realistic P intake, update p_intake = 90 in the animal_base.py file. The ration optimization and new animal requirements changes this summer will address this issue.

This is the branch in which I removed the duplicate animal lists, and now there is only one copy of all the animals in animal_management. I've also updated some commenting and code formatting.

### CODE REVIEW

File: RUFAS/routines/animal/clustering_pen_grouping.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/clustering_pen_grouping.py:66]: I've been meaning to ask you about this. How are protected/private class variables being used in animal?

**Review by wmaxdonovan — APPROVED** [review]
Looks great!

File: RUFAS/routines/animal/clustering_pen_grouping.py
Comment (militsasotirova) [reply RUFAS/routines/animal/clustering_pen_grouping.py:66]: I forgot that I had done this but a while ago I realized that we had protected class variables in our animal classes but we weren't treating them as such outside of those classes (we were accessing them as if they were public in files like animal_management.py and pen.py), so in this branch I went through and changed all of the variables to be public. Let me know if this is a problem, and I can change them back.

Comment (militsasotirova) [reply RUFAS/routines/animal/clustering_pen_grouping.py:66]: It was more of a consistency thing than anything else.

Comment (wmaxdonovan) [reply RUFAS/routines/animal/clustering_pen_grouping.py:66]: No, I think this is great.

File: RUFAS/routines/animal/life_cycle/animal_base.py
Comment (KFosterReed) [inline RUFAS/routines/animal/life_cycle/animal_base.py:71]: This may be a stupid question, but I don't understand the difference between init_from_animal and the above __init__ function. Can you explain why they are both required?

Comment (militsasotirova) [reply RUFAS/routines/animal/life_cycle/animal_base.py:71]: I believe that @katrina433 simplified this in her pull request so that there is only the __init__ function.

File: RUFAS/routines/animal/life_cycle/animal_base.py
Comment (KFosterReed) [inline RUFAS/routines/animal/life_cycle/animal_base.py]: I think that the 'prev' modification should go on the end, based on our naming conventions. I guess following that same line of thought, the 'd' for delta should also go on the end? Maybe we should check with Crop and Soil to see how they have used delta. @wmaxdonovan can you comment on this?

**Comment by KFosterReed** [comment]
First, I love the commenting. The link between the P-tracking pseudocode and the python files is very clear. Plus I like the highlighted TO DO comments and descriptions of where to make updates for ration and manure excretion. 

As far as I can tell the phosphorus tracking is working but there are some problems that I ran into that have to do with manure and grouping. The manure output is the same for all pens and doesn't seem to be connected to the number of animals in the pen. What's going on here? Is that connected to the fixed ration? 

Also, when I changed the pen sizes to see how that would affect the number of animals in each pen and possibly the manure per pen, I ran into an error. This is likely an issue for  @cvankerkhove. The error I got said it was an Unhandled exception: 9 and it was on L 123 of the clustering_pen_grouping.py file. 
    while percentile[i] <= index[key-1] or percentile[i] > index[key]:

I'd like to clear these things up before merging this branch, if possible.

**Comment by militsasotirova** [comment]
As far as the manure output goes - I looked into it and the same thing was happening for the ration. Each pen was correctly calculating its own total manure and total ration, however whenever one pen's manure and ration were calculated, all of the other pens were set to have those same values. Therefore, what you were seeing was the last pen's ration and manure values. I believe the issue had to do with how we were iterating over all of the pens to perform these calculations - I fixed this in my most recent commit.

**Comment by KFosterReed** [comment]
Great! Your fixes work and now the manure and ration track with the # of cows in the pen and are different per pen. I also realized that I was confused about which pens were for lactating cows and was able to change the pen sizes without getting the error so @cvankerkhove can ignore my last comment. We do need to make a note somewhere that the sum of the inputs for the lactating cow pen sizes need to be at least as big as the herd number to avoid an error. 

I have a few more things I want to look at but I hope to finish the review by Friday.

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (wmaxdonovan) [inline RUFAS/routines/animal/life_cycle/calf.py:53]: It looks like you changed a lot of these to public here already. Am I missing something / why are those above still private.

**Review by manfei-L — APPROVED** [review]
All new code matches the pseudocode for the p calculations, and I like how it fits in each animal class files. Great job!

File: RUFAS/routines/animal/life_cycle/calf.py
Comment (militsasotirova) [reply RUFAS/routines/animal/life_cycle/calf.py:53]: Which variables are you referring to?

**Comment by KFosterReed** [comment]
Alright! I am finally ready to approve this merge as well. Thanks to all for the great work :)

**Review by wmaxdonovan — APPROVED** [review]

**Pull request merged**
